Unpredictable, content based file names

erikrozendaal opened the pull request:
Instead of randomizing the file names they are now based on an HMAC of the generated content. The secret is generated randomly for every RRDP version and stored in the database.

Whenever the file is (re)created the filename is based on the HMAC SHA-256. So on re-creation the files will normally have the same name, unless the contents are now different due to code changes in the publication server.

erikrozendaal committed:
Always randomize file names when generating snapshots and deltas

Snapshot and delta files are supposed to be immutable and can be
cached "forever". However, when the publication server starts up it
re-generates the needed snapshot and delta files. But there is no
guarantee that the files will have exactly the same content (for
example, due to code changes or layout changes).

So now we always randomize the filenames and generate a new
notification.xml file.

This breaks multi-instance mode where multiple instances are running
from a single database, since the notification.xml file will not
match on every instance which may cause trouble when fetching the
referenced files from another instance.

The solution is to use deterministic filenames based on the contents,
such as using the SHA-256 of the file in the filename. However, since
we do not want to have predictable file names (due to caching of 404
responses) it should also include a secret and use an algorithm such
as HmacSHA256 to generate the unique filename.

This will be part of a later commit.

erikrozendaal committed:
Fix typo

erikrozendaal committed:
Make HashingSizedStream a proper Java OutputStream

erikrozendaal committed:
Base RRDP file names on content hash and version specific secret

Instead of fully randomizing the file names they are now based on an
HMAC of the generated content. The secret is generated randomly for
every RRDP version and stored in the database.

Whenever the file is (re)created the filename is based on the HMAC
SHA-256. On re-creation the files should have the same name, unless
the contents are now different due to code changes in the publication
server.

erikrozendaal committed:
Use Java 8 APIs

ties commented on src/main/scala/net/ripe/rpki/publicationserver/repository/HashingSizedStream.scala:
```diff
@@ -1,23 +1,33 @@
+  override def write(bytes: Array[Byte], offset: Int, len: Int): Unit = {
+    out.write(bytes, offset, len)
+    digest.update(bytes, offset, len)
+    mac.update(bytes, offset, len)
```
`out.write(bytes, offset, len)` and `digest.update` and `mac.update` are symmetrical w.r.t. behaviour -> this works.

ties commented on src/main/scala/net/ripe/rpki/publicationserver/repository/HashingSizedStream.scala:
```diff
@@ -1,23 +1,33 @@
+  }
 
-  def close() = os.close()
+  def summary = (Hash(digest.digest()), Bytes(mac.doFinal()), size)
```
`.doFinal` can only be called once.

Further calls to `write` should throw

erikrozendaal replied:
We could call `close` here (which is idempotent). This should cause I/O errors on subsequent `write` calls.

ties replied:
That sounds good. At least we will catch the error case early this way...

erikrozendaal committed:
Fix typo

Co-authored-by: Ties de Kock <ties@tiesdekock.nl>

erikrozendaal commented on src/main/resources/db/migration/R__functions.sql:
```diff
@@ -261,12 +262,12 @@ BEGIN
-        INSERT INTO versions (session_id, serial)
-        VALUES (uuid_in(md5(random()::TEXT || clock_timestamp()::TEXT)::CSTRING), 1)
+        INSERT INTO versions (session_id, serial, file_name_secret)
+        VALUES (uuid_in(md5(random()::TEXT || clock_timestamp()::TEXT)::CSTRING), 1, file_name_secret_)
```
We should be using `gen_random_uuid` which is part of the pgcrypto extension (and standard in PostgreSQL 13). Or pass it in from the JVM where we can securely generate the UUID without database extensions.

ties replied:
It looks like we are running PostgreSQL 12 at the moment - so perhaps using `sha256` would be fine. Let's ask if we can use 13 - I agree that `gen_random_uuid` is cleaner.

erikrozendaal committed:
Close stream before generating summary

After the summary is generated no writes should occur anymore. By
closing the stream write should result in an I/O exception by the
underlying stream, catching errors early.

lolepezy commented:
I guess, that's because I wasn't around during the discussion, but what's the git of this change? I mean, I see what it does, but _why_ do we want filenames to be not random, but content-based? It doesn't seem to simplify code a lot.

lolepezy commented:
> I guess, that's because I wasn't around during the discussion, but what's the aim of this change? I mean, I see what it does, but _why_ do we want filenames to be not random, but content-based? It doesn't seem to simplify code a lot.

I see, there's no more `knownName` concept, so the name is always known from the content, that  makes things easier indeed.
